File: activity/client/src/game/BonusManager.ts
import { BonusType, getBonusDef } from "./BonusType";
import { Element } from "./Element";
import type { LettreTour } from "./LettreTour";
import { RareteLettre } from "./RareteLettre";

export class BonusManager {
  private actifs: BonusType[] = [];
  private niveaux: Map<BonusType, number> = new Map();
  private globalMult: number = 1.0;

  activer(b: BonusType): boolean {
    if (this.actifs.includes(b) || this.actifs.length >= 3) return false;
    this.actifs.push(b);
    return true;
  }

  desactiver(b: BonusType): void {
    this.actifs = this.actifs.filter(x => x !== b);
  }

  setNiveau(b: BonusType, n: number): void {
    this.niveaux.set(b, n);
  }

  setGlobalMult(m: number): void { this.globalMult = m; }

  getActifs(): BonusType[] { return [...this.actifs]; }

  getMultiplicateur(tour: LettreTour, stat: string): number {
    return this.actifs.reduce((acc, b) => {
      const def = getBonusDef(b);
      if (!this.sApplique(b, tour)) return acc;
      if (def.statCiblee !== stat && def.statCiblee !== "*") return acc;
      const niveau = this.niveaux.get(b) ?? 1;
      // +10% of base buff per level above 1
      const effectiveMult = 1 + (def.mult - 1) * (1 + (niveau - 1) * 0.1);
      return acc * effectiveMult;
    }, 1.0) * this.globalMult;
  }

  private sApplique(b: BonusType, t: LettreTour): boolean {
    const ct = getBonusDef(b).cibleType;
    switch (ct) {
      case "CONSONNES": return !t.estVoyelle();
      case "VOYELLES":  return t.estVoyelle();
      case "FEU":       return t.element === Element.FEU;
      case "EAU":       return t.element === Element.EAU;
      case "TERRE":     return t.element === Element.TERRE;
      case "VENT":      return t.element === Element.VENT;
      case "RARES":     return t.rarete === RareteLettre.RARE;
      case "ALL":       return true;
      default:          return false;
    }
  }
}

File: src/BonusManager.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class BonusManager {

    private final List<BonusType> actifs = new ArrayList<>();

    /** Ajoute un bonus si absent et si le maximum de 3 n'est pas atteint. */
    public boolean activer(BonusType b) {
        if (actifs.contains(b) || actifs.size() >= 3) return false;
        actifs.add(b);
        return true;
    }

    public void desactiver(BonusType b) {
        actifs.remove(b);
    }

    public List<BonusType> getActifs() {
        return Collections.unmodifiableList(actifs);
    }

    /**
     * Retourne le produit des multiplicateurs de tous les bonus actifs
     * applicables à cette tour pour la stat demandée.
     */
    public double getMultiplicateur(LettreTour tour, String stat) {
        double result = 1.0;
        for (BonusType b : actifs) {
            boolean statMatch = b.getStatCiblee().equals(stat) || b.getStatCiblee().equals("*");
            if (statMatch && sApplique(b, tour)) {
                result *= b.getMultiplicateur();
            }
        }
        return result;
    }

    private boolean sApplique(BonusType b, LettreTour tour) {
        switch (b.getCibleType()) {
            case "CONSONNES": return !tour.estVoyelle();
            case "VOYELLES":  return tour.estVoyelle();
            case "FEU":       return tour.getElement() == Element.FEU;
            case "EAU":       return tour.getElement() == Element.EAU;
            case "TERRE":     return tour.getElement() == Element.TERRE;
            case "VENT":      return tour.getElement() == Element.VENT;
            case "RARES":     return tour.getRarete() == RareteLettre.RARE;
            case "ALL":       return true;
            default:          return false;
        }
    }
}
